File: src/game/sphereMath.ts
import * as THREE from 'three'

// 행성은 반지름 PLANET_RADIUS 의 구.
// 모든 캐릭터/장식은 구 표면(법선 방향 = 위쪽)에 정렬된다.
export const PLANET_RADIUS = 3

/** 위도(lat: -PI/2..PI/2), 경도(lon: -PI..PI) → 단위 방향 벡터 */
export function latLonToDir(lat: number, lon: number): THREE.Vector3 {
  const cosLat = Math.cos(lat)
  return new THREE.Vector3(
    cosLat * Math.sin(lon),
    Math.sin(lat),
    cosLat * Math.cos(lon),
  )
}

/** 단위 방향 벡터 → 위도/경도 */
export function dirToLatLon(dir: THREE.Vector3): { lat: number; lon: number } {
  const v = dir.clone().normalize()
  return { lat: Math.asin(THREE.MathUtils.clamp(v.y, -1, 1)), lon: Math.atan2(v.x, v.z) }
}

/**
 * 표면 위 한 점에서, 그 점의 법선(up)에 정렬되고 heading(라디안) 방향을 바라보는
 * 쿼터니언을 만든다. heading 0 = 북쪽(위도 증가 방향).
 */
export function surfaceQuaternion(dir: THREE.Vector3, heading: number): THREE.Quaternion {
  const up = dir.clone().normalize()
  // 임의의 기준에서 표면 접선 정북(north) 벡터 구하기
  const worldUp = new THREE.Vector3(0, 1, 0)
  let north = worldUp.clone().sub(up.clone().multiplyScalar(worldUp.dot(up)))
  if (north.lengthSq() < 1e-6) north = new THREE.Vector3(0, 0, 1) // 극점 예외
  north.normalize()
  const east = new THREE.Vector3().crossVectors(up, north).normalize()

  // heading 방향의 전방 벡터
  const forward = north.clone().multiplyScalar(Math.cos(heading))
    .add(east.clone().multiplyScalar(Math.sin(heading)))
    .normalize()

  const m = new THREE.Matrix4()
  // 모델은 +Z를 전방, +Y를 위로 가정
  const zAxis = forward
  const yAxis = up
  const xAxis = new THREE.Vector3().crossVectors(yAxis, zAxis).normalize()
  m.makeBasis(xAxis, yAxis, zAxis)
  return new THREE.Quaternion().setFromRotationMatrix(m)
}

/** 현재 방향에서 heading 방향으로 각거리 angle 만큼 표면을 따라 이동한 새 방향 */
export function moveOnSphere(dir: THREE.Vector3, heading: number, angle: number): THREE.Vector3 {
  const up = dir.clone().normalize()
  const worldUp = new THREE.Vector3(0, 1, 0)
  let north = worldUp.clone().sub(up.clone().multiplyScalar(worldUp.dot(up)))
  if (north.lengthSq() < 1e-6) north = new THREE.Vector3(0, 0, 1)
  north.normalize()
  const east = new THREE.Vector3().crossVectors(up, north).normalize()
  const forward = north.clone().multiplyScalar(Math.cos(heading))
    .add(east.clone().multiplyScalar(Math.sin(heading)))
    .normalize()
  // 회전축 = up × forward (오른손) → 표면을 따라 forward 방향으로 굴림
  const axis = new THREE.Vector3().crossVectors(forward, up).normalize()
  const q = new THREE.Quaternion().setFromAxisAngle(axis, angle)
  return up.clone().applyQuaternion(q).normalize()
}

File: unity/Assets/Scripts/SphereMath.cs
using UnityEngine;

namespace LittlePrince
{
    /// <summary>
    /// 구형 행성 위의 좌표/자세 수학. 웹 버전(src/game/sphereMath.ts)을
    /// Unity 좌표계(+Z 전방, +Y 위)로 이식.
    /// 행성은 원점 중심, 반지름 PlanetRadius 의 구. 표면 법선 = 위쪽.
    /// </summary>
    public static class SphereMath
    {
        public const float PlanetRadius = 3f;

        /// 위도(lat)·경도(lon, 라디안) → 단위 방향 벡터
        public static Vector3 LatLonToDir(float lat, float lon)
        {
            float cosLat = Mathf.Cos(lat);
            return new Vector3(cosLat * Mathf.Sin(lon), Mathf.Sin(lat), cosLat * Mathf.Cos(lon));
        }

        /// 단위 방향 벡터 → 위도·경도(라디안)
        public static void DirToLatLon(Vector3 dir, out float lat, out float lon)
        {
            Vector3 v = dir.normalized;
            lat = Mathf.Asin(Mathf.Clamp(v.y, -1f, 1f));
            lon = Mathf.Atan2(v.x, v.z);
        }

        /// 표면 한 점의 접선 기준틀: up(법선) / north(+Y극 방향 투영) / east
        public static void TangentBasis(Vector3 dir, out Vector3 up, out Vector3 north, out Vector3 east)
        {
            up = dir.normalized;
            Vector3 worldUp = Vector3.up;
            north = worldUp - up * Vector3.Dot(worldUp, up);
            if (north.sqrMagnitude < 1e-6f) north = Vector3.forward; // 극점 예외
            north.Normalize();
            east = Vector3.Cross(up, north).normalized;
        }

        /// heading(라디안, 0 = 북) 방향의 표면 접선 전방 벡터
        public static Vector3 HeadingForward(Vector3 dir, float heading)
        {
            TangentBasis(dir, out _, out Vector3 north, out Vector3 east);
            return (north * Mathf.Cos(heading) + east * Mathf.Sin(heading)).normalized;
        }

        /// 표면에 정렬되고 heading 을 바라보는 회전 (+Z 전방, +Y 위)
        public static Quaternion SurfaceRotation(Vector3 dir, float heading)
        {
            TangentBasis(dir, out Vector3 up, out _, out _);
            Vector3 forward = HeadingForward(dir, heading);
            return Quaternion.LookRotation(forward, up);
        }

        /// 현재 dir 에서 heading 방향으로 각거리 angle(라디안)만큼 표면 이동한 새 dir.
        /// up 과 forward 는 직교 → 두 벡터가 이루는 평면 내 회전으로 안전하게 계산
        /// (좌표계 손방향과 무관하게 항상 forward 쪽으로 이동).
        public static Vector3 MoveOnSphere(Vector3 dir, float heading, float angle)
        {
            Vector3 up = dir.normalized;
            Vector3 forward = HeadingForward(up, heading);
            return (up * Mathf.Cos(angle) + forward * Mathf.Sin(angle)).normalized;
        }
    }
}
